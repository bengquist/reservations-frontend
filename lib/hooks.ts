import { useEffect } from "react";

export function useClickAway(onClickAway: () => void, ref: any) {
  useEffect(() => {
    document.addEventListener("mousedown", handleClick);
    return () => {
      document.removeEventListener("mousedown", handleClick);
    };
  }, []);

  function handleClick(event: any) {
    if (ref.current && !ref.current.contains(event.target)) onClickAway();
  }
}
