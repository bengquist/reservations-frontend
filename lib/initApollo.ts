import { ApolloClient, InMemoryCache, HttpLink } from "apollo-boost";
import fetch from "isomorphic-unfetch";

let apolloClient: any = null;

function create(initialState: any) {
  return new ApolloClient({
    connectToDevTools: process.browser,
    ssrMode: !process.browser,
    link: new HttpLink({
      uri: "http://localhost:4000/graphql",
      credentials: "same-origin",
      fetch: !process.browser && fetch
    }),
    cache: new InMemoryCache().restore(initialState || {})
  });
}

export default function initApollo(initialState) {
  if (!process.browser) {
    return create(initialState);
  }

  if (!apolloClient) {
    apolloClient = create(initialState);
  }

  return apolloClient;
}
