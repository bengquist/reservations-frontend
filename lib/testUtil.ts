import casual from "casual";

casual.seed(333);

const mockReservation = () => ({
  arrivalDate: "06/09/2019",
  departureDate: "06/19/2019",
  hotelName: "Hilton Garden Inn",
  id: "j1h23489ydsa9sf8ha",
  name: "Blake"
});

const mockReservationList = () => [
  {
    arrivalDate: "06/09/2019",
    departureDate: "06/19/2019",
    hotelName: "Hilton Garden Inn",
    id: casual.uuid,
    name: "Blake"
  },
  {
    arrivalDate: "06/02/2019",
    departureDate: "06/10/2019",
    hotelName: "Homewood Suites",
    id: casual.uuid,
    name: "Ashlynn"
  },
  {
    arrivalDate: "06/05/2019",
    departureDate: "06/14/2019",
    hotelName: "The Statler Dallas",
    id: casual.uuid,
    name: "Tammy"
  },
  {
    arrivalDate: "06/11/2019",
    departureDate: "06/16/2019",
    hotelName: "Hampton Inn & Suites",
    id: casual.uuid,
    name: "Jessica"
  }
];

export { mockReservation, mockReservationList, casual };
