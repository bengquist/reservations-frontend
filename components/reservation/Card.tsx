import { Reservation } from "./types";
import styled, { css } from "styled-components";
import { hoverState, focusState } from "../styles/helpers";
import { formatDate } from "../../lib/formatHelpers";
import Modal from "../ui/Modal";
import Detail from "./Detail";
import { useState } from "react";

type Props = {
  reservation: Reservation;
};

const Card: React.FunctionComponent<Props> = ({ reservation }) => {
  const [showModal, setShowModal] = useState();

  return (
    <>
      <button onClick={() => setShowModal(true)} css={buttonStyle}>
        <Container>
          <span>{reservation.name}</span>
          <span>{reservation.hotelName}</span>
          <span className="date">{formatDate(reservation.arrivalDate)}</span>
          <span className="date">{formatDate(reservation.departureDate)}</span>
        </Container>
      </button>
      {showModal && (
        <Modal onClose={() => setShowModal(false)}>
          <Detail id={reservation.id} />
        </Modal>
      )}
    </>
  );
};

export default Card;

const buttonStyle = css`
  width: 100%;
  background: white;
  ${hoverState};
  ${focusState};
`;

const Container = styled.div`
  display: grid;
  grid-template-columns: 1.25fr 1.25fr 0.75fr 0.75fr;
  text-align: left;

  > span {
    padding: 1rem 0.5rem;
    border-right: 2px solid
      ${props => props.theme.colors && props.theme.colors.gray};
    :last-child {
      border-width: 1px 0 1px 1px;
    }
  }

  > .date {
    text-align: center;
  }

  @media (max-width: 500px) {
    font-size: 1rem;
  }
`;
