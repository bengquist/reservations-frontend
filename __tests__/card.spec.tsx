import * as React from "react";
import { shallow } from "enzyme";
import Card from "../components/reservation/Card";
import { formatDate } from "../lib/formatHelpers";
import { mockReservation } from "../lib/testUtil";

const setup = () => shallow(<Card reservation={mockReservation()} />);

describe("Card", () => {
  it("renders and matches snapshot", () => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("shows reservation name", () => {
    const wrapper = setup();

    const span = wrapper.find("span").at(0);

    expect(span.children().text()).toEqual(mockReservation().name);
  });

  it("shows hotel name", () => {
    const wrapper = setup();

    const span = wrapper.find("span").at(1);

    expect(span.children().text()).toEqual(mockReservation().hotelName);
  });

  it("shows arrival date", () => {
    const wrapper = setup();

    const span = wrapper.find("span").at(2);

    expect(span.children().text()).toEqual(
      formatDate(mockReservation().arrivalDate)
    );
  });

  it("shows departure date", () => {
    const wrapper = setup();

    const span = wrapper.find("span").at(3);

    expect(span.children().text()).toEqual(
      formatDate(mockReservation().departureDate)
    );
  });

  it("show modal on click", () => {
    const wrapper = setup();

    const button = wrapper.find("button");

    button.simulate("click");

    const modal = wrapper.find("Modal");

    expect(modal).not.toBeNull();
  });
});
