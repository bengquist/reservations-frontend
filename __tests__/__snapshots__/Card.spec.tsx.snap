// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders and matches snapshot 1`] = `
<Fragment>
  <button
    css={
      Array [
        "
  width: 100%;
  background: white;
  ",
        "
  transition: 0.2s;

  :hover,
  :active {
    background-color: ",
        [Function],
        ";
  }
",
        ";
  ",
        "
  transition: 0.1s;

  :focus {
    background-color: ",
        [Function],
        ";
    color: white;
  }
",
        ";
",
      ]
    }
    onClick={[Function]}
  >
    <styled.div>
      <span>
        Blake
      </span>
      <span>
        Hilton Garden Inn
      </span>
      <span
        className="date"
      >
        Jun 9th, 2019
      </span>
      <span
        className="date"
      >
        Jun 19th, 2019
      </span>
    </styled.div>
  </button>
</Fragment>
`;
